File: rust/crates/corlinman-tenant/src/pool.rs
//! `TenantPool` — multi-DB pool wrapper keyed by `(TenantId, db_name)`.
//!
//! Phase 2/3 each persistence module (`EvolutionStore`, `SqliteStore` for
//! kb, `SqliteSessionStore`, …) opened *one* `SqlitePool` against a
//! single fixed path. Phase 4 needs N tenants × M DB files, all opened
//! on demand and cached.
//!
//! `TenantPool` is intentionally **schema-agnostic**. It hands out raw
//! `SqlitePool`s; downstream crates clone-and-bind their own typed
//! repos against them. We do not centralise schema bootstrap here
//! because each crate already owns its `SCHEMA_SQL` + idempotent ALTER
//! block (`evolution::store::open` etc.) — duplicating that logic would
//! couple this crate to schema versions it has no business knowing.
//!
//! Concurrency:
//!   * the inner cache is a `Mutex<HashMap<...>>`, locked only for the
//!     "find or insert" step — pool open itself happens outside the
//!     lock so two simultaneous opens against *different* (tenant, db)
//!     pairs don't serialise on each other,
//!   * the open-race window for the *same* pair is closed by a
//!     "double-checked locking" pattern: re-check the map after the
//!     `connect_with` future resolves, and if another thread won the
//!     race we drop our pool and use theirs.

use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::str::FromStr;
use std::sync::Arc;
use std::time::Duration;

use sqlx::sqlite::{SqliteConnectOptions, SqliteJournalMode, SqlitePoolOptions, SqliteSynchronous};
use sqlx::SqlitePool;
use tokio::sync::Mutex;

use crate::id::TenantId;
use crate::path::{tenant_db_path, tenant_root_dir};

/// Errors emitted by [`TenantPool::get_or_open`].
#[derive(Debug, thiserror::Error)]
pub enum TenantPoolError {
    /// The tenant's directory could not be created (permission, full
    /// disk, parent missing, etc). Wrap rather than swallow so the
    /// gateway logs surface the underlying `io::Error::kind()`.
    #[error("create tenant dir {path}: {source}")]
    CreateDir {
        path: PathBuf,
        #[source]
        source: std::io::Error,
    },
    /// `SqliteConnectOptions::from_str` rejected the SQLite URL we
    /// derived from the path — usually means the path contains a
    /// character `sqlx` rejects in URLs (rare since `TenantId` is a
    /// slug, but file-name `name` is also caller-controlled).
    #[error("invalid sqlite url {url}: {source}")]
    InvalidUrl {
        url: String,
        #[source]
        source: sqlx::Error,
    },
    /// Pool open / first-connection failed. Surfaces SQLite's
    /// underlying error so a missing `pragma journal_mode` failure
    /// (read-only mount, etc) reaches the operator.
    #[error("connect sqlite {url}: {source}")]
    Connect {
        url: String,
        #[source]
        source: sqlx::Error,
    },
}

/// Composite key into the inner pool map. We keep this private so the
/// public API takes `(&TenantId, &str)` and doesn't leak the `String`
/// allocation.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
struct Key {
    tenant: TenantId,
    db_name: String,
}

/// Multi-DB pool wrapper. Cheap to clone (single `Arc` bump); every
/// clone shares the same underlying cache.
#[derive(Debug, Clone)]
pub struct TenantPool {
    /// Filesystem root holding `tenants/<tenant>/<name>.sqlite`. Always
    /// the gateway's `data_dir` in production; tempdirs in tests.
    root: Arc<Path>,
    inner: Arc<Mutex<HashMap<Key, SqlitePool>>>,
    /// Connection budget per `(tenant, db_name)` pool. Defaults to 8
    /// (matches `EvolutionStore::open` precedent). Per-pool cap rather
    /// than process-wide because contention bounds are also per-pool.
    max_connections: u32,
}

impl TenantPool {
    /// Build an empty pool wrapper rooted at `root`. Pools open
    /// lazily on first [`Self::get_or_open`] call; empty wrapper is
    /// cheap so tests / stripped-down builds don't pay a connection
    /// cost they won't use.
    pub fn new(root: impl AsRef<Path>) -> Self {
        Self {
            root: Arc::from(root.as_ref()),
            inner: Arc::new(Mutex::new(HashMap::new())),
            max_connections: 8,
        }
    }

    /// Override the per-pool connection cap (default 8). Builder
    /// pattern keeps the common `TenantPool::new(root)` form short.
    pub fn with_max_connections(mut self, n: u32) -> Self {
        self.max_connections = n;
        self
    }

    /// Filesystem root the wrapper was opened against. Tests use this
    /// to assert pool paths landed in the expected tempdir.
    pub fn root(&self) -> &Path {
        &self.root
    }

    /// Resolved path for `(tenant, db_name)`. Does **not** create the
    /// file — exposed so admin / migration code can probe whether the
    /// file already exists before deciding to open.
    pub fn db_path(&self, tenant: &TenantId, db_name: &str) -> PathBuf {
        tenant_db_path(&self.root, tenant, db_name)
    }

    /// Open (or return cached) `SqlitePool` for `(tenant, db_name)`.
    ///
    /// Creates the parent directory tree on first open so callers
    /// don't have to mkdir before invoking us. WAL +
    /// `synchronous=NORMAL` matches the rest of the codebase.
    pub async fn get_or_open(
        &self,
        tenant: &TenantId,
        db_name: &str,
    ) -> Result<SqlitePool, TenantPoolError> {
        let key = Key {
            tenant: tenant.clone(),
            db_name: db_name.to_owned(),
        };

        // Fast path: already cached.
        {
            let guard = self.inner.lock().await;
            if let Some(pool) = guard.get(&key) {
                return Ok(pool.clone());
            }
        }

        // Slow path: create dir + open pool *outside* the lock so two
        // first-time opens on different (tenant, db) pairs don't
        // serialise. We re-check inside the lock at the end for the
        // same-pair race.
        let dir = tenant_root_dir(&self.root, tenant);
        if let Err(source) = std::fs::create_dir_all(&dir) {
            return Err(TenantPoolError::CreateDir { path: dir, source });
        }

        let path = tenant_db_path(&self.root, tenant, db_name);
        let url = format!("sqlite://{}", path.display());
        let options = SqliteConnectOptions::from_str(&url)
            .map_err(|source| TenantPoolError::InvalidUrl {
                url: url.clone(),
                source,
            })?
            .create_if_missing(true)
            .journal_mode(SqliteJournalMode::Wal)
            .synchronous(SqliteSynchronous::Normal)
            .foreign_keys(true)
            .busy_timeout(Duration::from_secs(5));

        let pool = SqlitePoolOptions::new()
            .max_connections(self.max_connections)
            .connect_with(options)
            .await
            .map_err(|source| TenantPoolError::Connect { url, source })?;

        // Same-pair race close: another caller may have raced us in.
        // If so, drop our pool and use theirs so we don't leak file
        // descriptors or end up with two pools fighting for the same
        // WAL file.
        let mut guard = self.inner.lock().await;
        if let Some(existing) = guard.get(&key) {
            drop(pool);
            return Ok(existing.clone());
        }
        guard.insert(key, pool.clone());
        Ok(pool)
    }

    /// True iff the pool for `(tenant, db_name)` is already cached.
    /// Tests use this to assert lazy-open behaviour without forcing a
    /// connection. Public so the gateway's `tenant create` path can
    /// short-circuit a re-open after migrations.
    pub async fn is_cached(&self, tenant: &TenantId, db_name: &str) -> bool {
        let key = Key {
            tenant: tenant.clone(),
            db_name: db_name.to_owned(),
        };
        let guard = self.inner.lock().await;
        guard.contains_key(&key)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::TempDir;

    fn acme() -> TenantId {
        TenantId::new("acme").unwrap()
    }

    fn bravo() -> TenantId {
        TenantId::new("bravo").unwrap()
    }

    #[tokio::test]
    async fn opens_pool_lazily_and_caches() {
        let tmp = TempDir::new().unwrap();
        let pool = TenantPool::new(tmp.path());

        let tenant = acme();

        assert!(!pool.is_cached(&tenant, "evolution").await);

        let p1 = pool.get_or_open(&tenant, "evolution").await.unwrap();
        assert!(pool.is_cached(&tenant, "evolution").await);

        let p2 = pool.get_or_open(&tenant, "evolution").await.unwrap();
        // SqlitePool wraps an Arc — two clones of the same underlying
        // pool share state; we sanity-check by issuing a query through
        // both.
        sqlx::query("SELECT 1").execute(&p1).await.unwrap();
        sqlx::query("SELECT 1").execute(&p2).await.unwrap();
    }

    #[tokio::test]
    async fn isolates_pools_per_tenant() {
        let tmp = TempDir::new().unwrap();
        let pool = TenantPool::new(tmp.path());

        // Each tenant gets its own DB file.
        let _a = pool.get_or_open(&acme(), "evolution").await.unwrap();
        let _b = pool.get_or_open(&bravo(), "evolution").await.unwrap();

        let p_acme = tmp.path().join("tenants/acme/evolution.sqlite");
        let p_bravo = tmp.path().join("tenants/bravo/evolution.sqlite");
        assert!(p_acme.exists(), "acme db should exist at {p_acme:?}");
        assert!(p_bravo.exists(), "bravo db should exist at {p_bravo:?}");

        // Distinct files: writing to one does not appear in the other.
        let acme_pool = pool.get_or_open(&acme(), "evolution").await.unwrap();
        sqlx::query("CREATE TABLE marker (x INTEGER)")
            .execute(&acme_pool)
            .await
            .unwrap();
        let bravo_pool = pool.get_or_open(&bravo(), "evolution").await.unwrap();
        // The marker table is acme-only; sqlite_master on bravo must
        // not contain it.
        let row: Option<(String,)> =
            sqlx::query_as("SELECT name FROM sqlite_master WHERE type='table' AND name='marker'")
                .fetch_optional(&bravo_pool)
                .await
                .unwrap();
        assert!(row.is_none(), "bravo db must not see acme marker table");
    }

    #[tokio::test]
    async fn isolates_pools_per_db_name() {
        let tmp = TempDir::new().unwrap();
        let pool = TenantPool::new(tmp.path());
        let tenant = acme();

        let _evol = pool.get_or_open(&tenant, "evolution").await.unwrap();
        let _kb = pool.get_or_open(&tenant, "kb").await.unwrap();

        assert!(tmp.path().join("tenants/acme/evolution.sqlite").exists());
        assert!(tmp.path().join("tenants/acme/kb.sqlite").exists());
    }

    #[tokio::test]
    async fn db_path_does_not_create_file() {
        let tmp = TempDir::new().unwrap();
        let pool = TenantPool::new(tmp.path());
        let tenant = acme();

        // Probing the path is allowed before open; it must not touch
        // the filesystem (admin / migration code relies on this).
        let p = pool.db_path(&tenant, "evolution");
        assert_eq!(p, tmp.path().join("tenants/acme/evolution.sqlite"));
        assert!(!p.exists());
    }

    #[tokio::test]
    async fn concurrent_first_open_does_not_panic_or_deadlock() {
        // Two tasks racing on the same (tenant, db) pair. The
        // double-checked-lock pattern in `get_or_open` should make one
        // win and the other observe the cached pool.
        let tmp = TempDir::new().unwrap();
        let pool = TenantPool::new(tmp.path());
        let p1 = pool.clone();
        let p2 = pool.clone();
        let t = acme();
        let t1 = t.clone();
        let t2 = t.clone();

        let h1 = tokio::spawn(async move { p1.get_or_open(&t1, "evolution").await.is_ok() });
        let h2 = tokio::spawn(async move { p2.get_or_open(&t2, "evolution").await.is_ok() });
        assert!(h1.await.unwrap());
        assert!(h2.await.unwrap());

        // Exactly one cached entry.
        assert!(pool.is_cached(&t, "evolution").await);
    }
}

File: python/packages/corlinman-server/src/corlinman_server/tenancy/pool.py
"""``TenantPool`` — multi-DB pool wrapper keyed by ``(TenantId, db_name)``.

Python port of ``corlinman-tenant::pool``. The Rust crate hands out
``sqlx::SqlitePool`` (a connection pool); :mod:`aiosqlite` exposes
single connections, not pools, so the Python port stores one cached
:class:`aiosqlite.Connection` per ``(tenant, db_name)`` pair and serves
that to callers — preserving the "lazy-open + cache + share" contract
even though the underlying primitive differs.

Pragmas applied at open match the Rust side:

* ``journal_mode=WAL`` — concurrent readers + a single writer,
* ``synchronous=NORMAL`` — durability on power-loss, but skips fsync on
  every commit (the chosen point on the durability/throughput curve for
  the rest of the corlinman SQLite stores),
* ``foreign_keys=ON`` — enforced by SQLite per-connection,
* ``busy_timeout=5000`` — 5-second wait on contention before raising.

Concurrency: the cache is guarded by an :class:`asyncio.Lock`, locked
only for the "find or insert" step. The first-open of a single
``(tenant, db_name)`` pair runs under the lock so two concurrent opens
of the same file don't race in WAL setup — exactly the Phase 4 W1.5
/A7 flake fix the Rust side ships.
"""

from __future__ import annotations

import asyncio
import contextlib
from pathlib import Path
from typing import NamedTuple

import aiosqlite

from corlinman_server.tenancy.id import TenantId
from corlinman_server.tenancy.path import tenant_db_path, tenant_root_dir


class TenantPoolError(RuntimeError):
    """Base class for :meth:`TenantPool.get_or_open` errors."""


class TenantPoolCreateDirError(TenantPoolError):
    """The tenant's directory could not be created (permission, full
    disk, parent missing, etc). Wraps the underlying
    :class:`OSError`."""

    def __init__(self, path: Path, source: OSError) -> None:
        self.path = path
        self.source = source
        super().__init__(f"create tenant dir {path}: {source}")


class TenantPoolConnectError(TenantPoolError):
    """Connection open or initial pragma application failed. Wraps the
    underlying :class:`aiosqlite.Error`."""

    def __init__(self, db_path: Path, source: BaseException) -> None:
        self.db_path = db_path
        self.source = source
        super().__init__(f"connect sqlite {db_path}: {source}")


class _Key(NamedTuple):
    """Composite cache key. Private — the public API takes
    ``(TenantId, str)`` and we materialise the key internally so the
    surface doesn't leak the tuple shape."""

    tenant: TenantId
    db_name: str


class TenantPool:
    """Multi-DB connection wrapper. Cheap to share across coroutines;
    every instance shares its own cache via the internal lock and
    dict — clone semantics from the Rust ``Arc`` version are not
    needed in Python (just hand the instance around).

    Connections opened through :meth:`get_or_open` are cached forever
    (process lifetime) unless explicitly closed via :meth:`close_all`.
    """

    # Connection budget per `(tenant, db_name)` cached entry. The Rust
    # side allocates an N-connection pool here (default 8); aiosqlite
    # serves a single async connection per file, so the cap is "1
    # connection per pair". Kept as an attribute for parity with the
    # builder API and so a future port to a real pool (e.g.
    # ``sqlalchemy.ext.asyncio``) can flip it.
    _DEFAULT_MAX_CONNECTIONS = 8

    def __init__(self, root: Path | str) -> None:
        """Build an empty wrapper rooted at ``root``. Connections
        open lazily on first :meth:`get_or_open` — an empty wrapper is
        cheap so tests / stripped-down builds don't pay any I/O they
        won't use."""
        self._root: Path = Path(root)
        self._cache: dict[_Key, aiosqlite.Connection] = {}
        self._lock = asyncio.Lock()
        self._max_connections = self._DEFAULT_MAX_CONNECTIONS

    # ---- builders --------------------------------------------------------------

    def with_max_connections(self, n: int) -> TenantPool:
        """Override the per-pool connection cap (default 8). Mirrors the
        Rust builder; in the aiosqlite port the value is stored for
        parity but not yet enforced (one connection per pair is the
        only shape aiosqlite supports without a pool layer)."""
        self._max_connections = n
        return self

    # ---- accessors -------------------------------------------------------------

    def root(self) -> Path:
        """Filesystem root the wrapper was opened against. Tests use
        this to assert connection paths landed in the expected tempdir."""
        return self._root

    def db_path(self, tenant: TenantId, db_name: str) -> Path:
        """Resolved path for ``(tenant, db_name)``. Does **not** create
        the file — exposed so admin / migration code can probe whether
        the file already exists before deciding to open."""
        return tenant_db_path(self._root, tenant, db_name)

    async def is_cached(self, tenant: TenantId, db_name: str) -> bool:
        """True iff the connection for ``(tenant, db_name)`` is already
        cached. Tests use this to assert lazy-open behaviour without
        forcing a connection. Public so the gateway's ``tenant create``
        path can short-circuit a re-open after migrations."""
        key = _Key(tenant, db_name)
        async with self._lock:
            return key in self._cache

    # ---- main entrypoint -------------------------------------------------------

    async def get_or_open(self, tenant: TenantId, db_name: str) -> aiosqlite.Connection:
        """Open (or return cached) :class:`aiosqlite.Connection` for
        ``(tenant, db_name)``.

        Creates the parent directory tree on first open so callers
        don't have to mkdir before invoking us. WAL +
        ``synchronous=NORMAL`` + ``foreign_keys=ON`` matches the rest
        of the codebase.

        Raises:
          * :class:`TenantPoolCreateDirError` — directory create failed.
          * :class:`TenantPoolConnectError` — connection open / pragma
            apply failed.
        """
        key = _Key(tenant, db_name)

        # Fast path: already cached. Held briefly to avoid blocking
        # unrelated opens.
        async with self._lock:
            cached = self._cache.get(key)
            if cached is not None:
                return cached

            # Slow path: serialise open of this (tenant, db) pair under
            # the lock. Earlier Rust revisions opened *outside* the
            # lock to keep distinct-pair opens parallel, but Phase 4
            # W1.5 /A7 caught a flake where two simultaneous SQLite
            # opens of the *same* file race in `create_if_missing` +
            # WAL setup and one transiently errors. The fix —
            # serialise — also simplifies the Python port (one
            # `asyncio.Lock`, no double-checked locking, no risk of
            # awaiting `connect` outside the lock and then losing the
            # race anyway).
            dir_path = tenant_root_dir(self._root, tenant)
            try:
                dir_path.mkdir(parents=True, exist_ok=True)
            except OSError as exc:
                raise TenantPoolCreateDirError(dir_path, exc) from exc

            file_path = tenant_db_path(self._root, tenant, db_name)
            try:
                conn = await aiosqlite.connect(str(file_path))
                # Pragmas mirror the Rust SqliteConnectOptions chain.
                # WAL + NORMAL is the codebase default; foreign_keys
                # is per-connection so we set it on every open;
                # busy_timeout absorbs short contention windows
                # without surfacing SQLITE_BUSY to the caller.
                await conn.execute("PRAGMA journal_mode=WAL")
                await conn.execute("PRAGMA synchronous=NORMAL")
                await conn.execute("PRAGMA foreign_keys=ON")
                await conn.execute("PRAGMA busy_timeout=5000")
                # Default row factory: aiosqlite returns tuples; the
                # admin_schema layer accesses columns positionally,
                # so we leave the factory alone here. Callers that
                # want dict-style access can swap `conn.row_factory`
                # at the call site.
            except BaseException as exc:
                # Cover both aiosqlite.Error and the underlying
                # sqlite3.OperationalError surfaces — wrap them so
                # callers don't have to import the aiosqlite tree.
                raise TenantPoolConnectError(file_path, exc) from exc

            self._cache[key] = conn
            return conn

    # ---- lifecycle -------------------------------------------------------------

    async def close_all(self) -> None:
        """Close every cached connection and drop the cache. Idempotent
        — calling again on an already-drained pool is a no-op. Used at
        process shutdown so SQLite's WAL files get checkpointed
        cleanly; tests that want to assert "no connections leaked"
        also call this in teardown."""
        async with self._lock:
            connections = list(self._cache.values())
            self._cache.clear()
        # Close outside the lock so a slow close (WAL checkpoint, etc.)
        # doesn't keep new opens waiting. Surfacing one bad connection
        # close shouldn't mask the others — the Rust side relies on
        # sqlx's drop to clean up; the Python equivalent is best-effort.
        for conn in connections:
            with contextlib.suppress(Exception):
                await conn.close()


__all__ = [
    "TenantPool",
    "TenantPoolConnectError",
    "TenantPoolCreateDirError",
    "TenantPoolError",
]
